### ISSUE TITLE
provisioning: CATTLE_ADDRESS and CATTLE_INTERNAL_ADDRESS Agent Environment Vars are not supported

### ISSUE THREAD

**Issue by fgiudici** [rancher/elemental-operator#841]
From a user report:
Installing RKE2 via the Rancher UI, the Agent Environment Vars CATTLE_ADDRESS and CATTLE_INTERNAL_ADDRESS are not honored when building an Elemental Cluster.

The same variables are honored when building a Custom Cluster instead.

Investigation needed.

**Comment by fgiudici** [comment]
Note: HTTP_PROXY and HTTPS_PROXY Agent Environment Vars are honored instead.

**Comment by fgiudici** [comment]
The rancher-system-agent.sh script is started from an elemental-system-agent plan.
Trying to reproduce the issue and following the single steps, the script is correctly embedded in the target machine secret for consumption from the elemental-system-agent.

The script has "embedded" all the specified environment variables, included the CATTLE_ADDRESS and CATTLE_INTERNAL_ADDRESS ones.

All looks correct, no issue found.

Some more investigation required in collaboration with the user.

**Comment by DCSyd** [comment]
In case of custom cluster 50-rancher.yaml gets update with internal and external IP addresses but not in case of elemental

```
#cat /etc/rancher/rke2/config.yaml.d/50-rancher.yaml 
{
  "advertise-address": "192.168.100.65",
  "agent-token": "t5wv6xgfljl8h76f2ch75snvb87chnznkhmrgr7fpckb7h8qkl8gzl",
  "cni": "calico",
  "disable-kube-proxy": false,
  "etcd-expose-metrics": false,
  "etcd-snapshot-retention": 5,
  "etcd-snapshot-schedule-cron": "0 */5 * * *",
  "kube-controller-manager-arg": [
    "cert-dir=/var/lib/rancher/rke2/server/tls/kube-controller-manager",
    "secure-port=10257"
  ],
  "kube-controller-manager-extra-mount": [
    "/var/lib/rancher/rke2/server/tls/kube-controller-manager:/var/lib/rancher/rke2/server/tls/kube-controller-manager"
  ],
  "kube-scheduler-arg": [
    "cert-dir=/var/lib/rancher/rke2/server/tls/kube-scheduler",
    "secure-port=10259"
  ],
  "kube-scheduler-extra-mount": [
    "/var/lib/rancher/rke2/server/tls/kube-scheduler:/var/lib/rancher/rke2/server/tls/kube-scheduler"
  ],
  **_"node-external-ip": [
    "192.168.8.140"
  ],
  "node-ip": [
    "192.168.100.65"
  ],_**
  "node-label": [
    "cattle.io/os=linux",
    "rke.cattle.io/machine=6704ca34-385c-411e-85b1-77958341e0f2"
  ],
  "private-registry": "/etc/rancher/rke2/registries.yaml",
  "protect-kernel-defaults": false,
  "system-default-registry": "registry.rancher.com",
  "tls-san": [
    "192.168.8.140"
  ],
  "token":
```

**Comment by DCSyd** [comment]
```
 "node-external-ip": [
    "192.168.8.140"
  ],
  "node-ip": [
    "192.168.100.65"
  ],
```

**Comment by fgiudici** [comment]
TL;DR: Rancher Provisioning v2 currently does not support CATTLE_ADDRESS and CATTLE_INTERNAL_ADDRESS agent variables with Elemental provisioning. There is no easy workaround to address this.

Long Version
Elemental leverages the Rancher provisioning v2 for k3s/RKE2 deployment.
While this is the same Rancher provisioning functionality used by the "Custom" k3s/RKE2 Rancher Cluster creation, Elemental exposes the MachineInventorySelector resource that acts as a Cluster API Infrastructure Provider for the Rancher CAPI functionality to automate the rancher-system-installation.sh script execution.
Long story short, this impacts on what, how and when the required kubernetes resources needed for Cluster creation are created by Rancher.
In particular, the machine-plan secrets consumed by the rancher-system-agent of each enrolled host, are created at Cluster creation for Elemental, while are created on-the-fly for the host of "Custom" Clusters (when the rancher-system-install.sh script is called with the one-liner suggested by the Rancher UI).
The k3s/RKE2 installation and the creation of the related yaml configuration (`config.yaml.d/50-rancher.yaml ` file) are performed through these machine-plan secret we just mentioned.
The  k3s/RKE2 `node-ip` and `node-external-ip` sections in the `config.yaml.d/50-rancher.yaml` file are filled by Rancher provisioning from the `rke.cattle.io/internal-address` and `rke.cattle.io/address` annotations in the machine-plan secret (when present).
When running the rancher-system-install.sh script, the CATTLE_ADDRESS and CATTLE_INTERNAL_ADDRESS variables are:
1) converted to actual IP addresses (when they refer an interface name)
2) passed as headers in the v3/connect/agent Rancher provisioning API 
v3/connect/agent Rancher provisioning will check if the host invoking the script has already a machine-plan secret associated: if not, it will create a new one, adding the `rke.cattle.io/internal-address` and `rke.cattle.io/address` annotations filled with the ip addresses from the CATTLE_ADDRESS and CATTLE_INTERNAL_ADDRESS variables.
In case of Elemental, as we said, the machine-plan secrets are created at Cluster creation time, so while the Elemental host invokes the system-agent-install.sh script with the right CATTLE_ADDRESS and CATTLE_INTERNAL_ADDRESS variables, Rancher will find an already existing machine-plan secret for the host and will not update it with the  `rke.cattle.io/internal-address` and `rke.cattle.io/address`  annotations.
When later the machine-plan secret's plan is filled by Rancher, since the `rke.cattle.io/internal-address` and `rke.cattle.io/address`  annotations are missing, the  `config.yaml.d/50-rancher.yaml` file will miss the "node-ip" and the "node-external-ip" sections.

In order to get the "node-ip" and "node-external-ip" section automatically, one should annotate each machine-plan secret with the `rke.cattle.io/internal-address` and `rke.cattle.io/address`  keys before the host provisioning starts. Anyway, the values there should be  IP addresses and not an interface names (the logic to convert from interface name to ip address is in the system-agent-install.sh  script and applied before writing the annotations).
This makes a workaround impractical.

This issue could be address either in Rancher provisioning (annotating the machine-plan secrets when needed) or in the Elemental Operator (patching the machine-plan secrets itself or patching the k3s/RKE2 configuration directly).
